src/interactive/completions.ts

```diff
@@ -36,42 +37,34 @@ const requestTypeGetCompletionItems = new rpc.RequestType<
     void
     >('repl/getcompletions')
 
+
 function completionItemProvider(conn: MessageConnection): vscode.CompletionItemProvider {
     return {
         provideCompletionItems: async (document, position, token, context) => {
             if (!vscode.workspace.getConfiguration('julia.runtimeCompletions')) {
                 return
             }
-            const completionPromise = (async () => {
-                const startPosition = new vscode.Position(position.line, 0)
-                const lineRange = new vscode.Range(startPosition, position)
-                const line = document.getText(lineRange)
-                const mod: string = await getModuleForEditor(document, position)
-                return {
-                    items: await conn.sendRequest(requestTypeGetCompletionItems, { line, mod }),
-                    isIncomplete: true
-                }
-            })()
 
-            const cancelPromise: Promise<vscode.CompletionList> = new Promise(resolve => {
-                token.onCancellationRequested(() => resolve({
-                    items: [],
-                    isIncomplete: true
-                }))
-                setTimeout(() => {
-                    if (!token.isCancellationRequested) {
-                        resolve({
-                            items: [],
-                            isIncomplete: true
-                        })
-                    }
-                }, 500)
-            })
+            const newToken = combineCancellationTokens(
+                token,
+                createTimeoutCancellation(500)
+            )
+
+            const startPosition = new vscode.Position(position.line, 0)
+            const lineRange = new vscode.Range(startPosition, position)
+            const line = document.getText(lineRange)
+            const mod: string = await getModuleForEditor(document, position) // TODO pass newToken once we support tokens in JSONRPC.jl
+
+            if (newToken.isCancellationRequested) { return }
 
-            return Promise.race([
-                completionPromise,
-                cancelPromise
-            ])
+            const items = await conn.sendRequest(requestTypeGetCompletionItems, { line, mod }) // TODO pass newToken once we support tokens in JSONRPC.jl
```

Comment:
Doesn't this mean that we always wait for this to go through before returning, unless cancellation got requested *before* we dispatch this request?

Reply:
Ah, yes, you're right. I don't know actually what the semantics are if we were to pass the token to the `sendRequest` call: would this `await` here finish right away, irrespective of what is happening on the server? In that case we could just pass the token here, even if the server doesn't yet implement cancel notifications. Or would it send the cancel notification message to the server and then wait until the request returns?

Reply:
The latter, I think.

Reply:
Yep, seems more likely...